graphql/src/main/java/com/linecorp/armeria/server/graphql/DefaultGraphQLService.java

```diff
@@ -0,0 +1,220 @@
+            return HttpResponse.of(HttpStatus.UNSUPPORTED_MEDIA_TYPE,
+                                   MediaType.PLAIN_TEXT,
+                                   "Could not process GraphQL request");
+        }
+        if (MediaType.isJson(contentType)) {
+            return HttpResponse.from(request.aggregate().handle((req, thrown) -> {
+                if (thrown != null) {
+                    return HttpResponse.ofFailure(thrown);
+                }
+                final String body = req.contentUtf8();
+                if (Strings.isNullOrEmpty(body)) {
+                    return HttpResponse.of(HttpStatus.BAD_REQUEST, MediaType.PLAIN_TEXT, "Body is required");
+                }
+                final Map<String, Object> requestMap = parseJsonString(body);
+                final String query = (String) requestMap.getOrDefault("query", "");
+                return execute(ctx, executionInput(query, ctx,
+                                                   toVariableMap(requestMap.get("variables")),
+                                                   (String) requestMap.get("operationName")));
+            }));
+        }
+        if (contentType.is(MediaType.GRAPHQL)) {
+            return HttpResponse.from(request.aggregate().handle((req, thrown) -> {
+                if (thrown != null) {
+                    return HttpResponse.ofFailure(thrown);
+                }
+                final String query = req.contentUtf8();
+                return execute(ctx, executionInput(query, ctx, null, null));
+            }));
+        }
+        return HttpResponse.of(HttpStatus.UNSUPPORTED_MEDIA_TYPE,
+                               MediaType.PLAIN_TEXT,
+                               "Could not process GraphQL request");
+    }
+
+    private ExecutionInput executionInput(String query, ServiceRequestContext ctx,
+                                          @Nullable Map<String, Object> variables,
+                                          @Nullable String operationName) {
+        final Builder builder = ExecutionInput.newExecutionInput(query);
+        if (variables != null && !variables.isEmpty()) {
+            builder.variables(variables);
+        }
+        if (operationName != null) {
+            builder.operationName(operationName);
+        }
+        if (dataLoaderRegistry != null) {
+            builder.dataLoaderRegistry(dataLoaderRegistry);
+        }
+        return builder.context(ctx).build();
+    }
+
+    private HttpResponse execute(ServiceRequestContext ctx, ExecutionInput input) {
+        if (useBlockingTaskExecutor) {
+            final CompletableFuture<HttpResponse> future = new CompletableFuture<>();
+            ctx.blockingTaskExecutor().execute(() -> {
+                final ExecutionResult executionResult = graphQL.execute(input);
+                future.complete(HttpResponse.of(MediaType.JSON_UTF_8,
+                                                toJsonString(executionResult.toSpecification())));
```

Comment:
When `toJsonString` raises an exception the response is not sent. Could you fix that please? 😄

Reply:
Hmm... I'll think about it. 🧐